x-pack/plugins/lens/public/datatable_visualization/components/columns.tsx

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React from 'react';
+import { i18n } from '@kbn/i18n';
+import { EuiDataGridColumn, EuiDataGridColumnCellActionProps } from '@elastic/eui';
+import type { Datatable, DatatableColumnMeta } from 'src/plugins/expressions';
+import type { FormatFactory } from '../../types';
+import type { DatatableColumns } from './types';
+
+export const createGridColumns = (
+  bucketColumns: string[],
+  table: Datatable,
+  handleFilterClick: (
+    field: string,
+    value: unknown,
+    colIndex: number,
+    rowIndex: number,
+    negate?: boolean
+  ) => void,
+  isReadOnly: boolean,
+  columnConfig: DatatableColumns & { type: 'lens_datatable_columns' },
+  visibleColumns: string[],
+  formatFactory: FormatFactory,
+  onColumnResize: (eventData: { columnId: string; width: number | undefined }) => void
+) => {
+  const columnsReverseLookup = table.columns.reduce<
+    Record<string, { name: string; index: number; meta?: DatatableColumnMeta }>
+  >((memo, { id, name, meta }, i) => {
+    memo[id] = { name, index: i, meta };
+    return memo;
+  }, {});
+
+  const bucketLookup = new Set(bucketColumns);
+
+  const getContentData = ({
+    rowIndex,
+    columnId,
+  }: Pick<EuiDataGridColumnCellActionProps, 'rowIndex' | 'columnId'>) => {
+    const rowValue = table.rows[rowIndex][columnId];
+    const column = columnsReverseLookup[columnId];
+    const contentsIsDefined = rowValue != null;
+
+    const cellContent = formatFactory(column?.meta?.params).convert(rowValue);
+    return { rowValue, contentsIsDefined, cellContent };
+  };
+
+  return visibleColumns.map((field) => {
+    const filterable = bucketLookup.has(field);
+    const { name, index: colIndex } = columnsReverseLookup[field];
+
+    const cellActions = filterable
+      ? [
+          ({ rowIndex, columnId, Component, closePopover }: EuiDataGridColumnCellActionProps) => {
+            const { rowValue, contentsIsDefined, cellContent } = getContentData({
+              rowIndex,
+              columnId,
+            });
+
+            const filterForText = i18n.translate(
```

Comment:
This code is being executed for every bucket cell, so it's a lot. And it's really easy to move outside of the `createGridColumns` function, as `filterForText` and `filterForAriaLabel(cellContent)`. What do you think about it?

Reply:
Mind the `AriaLabel` requires some contextual value

Reply:
After reviewing the code, actually I'm not so sure about moving it: while it is true that each cell will have these actions, it seems to me that only one cell per time will execute the code in the scope of the function. So no much a performance trouble I'd say.